### PR TITLE
Feature/add checkers

### PR DESCRIPTION
**Pull request by DavidSubiros**
### What

- Moved S3 interactions to dp-s3, and use them here.
- Use new libraries with health checks, instead of go-ns.
- Zebedee health client only used if we are in publishing mode.
- Registered Checkers for health check.
- Use latest health check with correct status code according to global health status.
Note: I created a new struct called Uploader instead of using the S3 client to match the AWS SDK and Cryptoclient code structure.

### How to review

- Unit tests should pass.
- Check that code changes make sense.
- You can try the health endpoint with and without running dependencies.

### Who can review

Anyone.

### CODE REVIEW

File: service/service.go [hunk 1]
```diff
@@ -87,7 +71,7 @@ func Create(
 	// For non-whitelisted endpoints, do identityHandler or corsHandler
 	if cfg.IsPublishing {
 		log.Event(ctx, "private endpoints are enabled. using identity middleware")
-		identityHandler := identity.Handler(cfg.ZebedeeURL)
+		identityHandler := identity.HandlerForHTTPClient(clientsidentity.NewAPIClient(zc.Client, cfg.ZebedeeURL))
```
Comment (DavidSubiros) [inline service/service.go]: We might want to use dp-api-clients-go identity here, instead of go-ns/identity

**Comment by DavidSubiros** [comment]
Adding severity to logs before merging this PR

**Comment by CarlHembrough** [comment]
I have run an end to end test on this and everything appears to be working fine. When looking at the healthcheck logs it appears that the dataset-api and filter-api are checked twice? Any ideas why this is happening? It doesn't seem to be the case for Zebedee:

```
{
  "created_at": "2020-03-16T09:20:59.078182Z",
  "data": {
    "action": "retrieving dataset",
    "method": "GET",
    "uri": "http://localhost:22000"
  },
  "event": "Making request to service: dataset-api",
  "namespace": "dp-download-service",
  "severity": 3
}
{
  "created_at": "2020-03-16T09:20:59.086234Z",
  "data": {
    "action": "retrieving dataset",
    "method": "GET",
    "uri": "http://localhost:22000"
  },
  "event": "Making request to service: dataset-api",
  "namespace": "dp-download-service",
  "severity": 3
}
{
  "created_at": "2020-03-16T09:20:59.390128Z",
  "data": {
    "action": "retrieving dataset",
    "method": "GET",
    "uri": "http://localhost:8082"
  },
  "event": "Making request to service: Zebedee",
  "namespace": "dp-download-service",
  "severity": 3
}
{
  "created_at": "2020-03-16T09:21:00.439796Z",
  "data": {
    "action": "retrieving dataset",
    "method": "GET",
    "uri": "http://localhost:22100"
  },
  "event": "Making request to service: filter-api",
  "namespace": "dp-download-service",
  "severity": 3
}
{
  "created_at": "2020-03-16T09:21:00.445315Z",
  "data": {
    "action": "retrieving dataset",
    "method": "GET",
    "uri": "http://localhost:22100"
  },
  "event": "Making request to service: filter-api",
  "namespace": "dp-download-service",
  "severity": 3
}
```

File: service/service.go [hunk 2]
```diff
@@ -7,69 +7,54 @@ import (
...
 
 	gorillahandlers "github.com/gorilla/handlers"
 )
 
 // Download represents the configuration to run the download service
 type Download struct {
-	datasetClient DatasetClient
-	filterClient  FilterClient
-	vaultClient   VaultClient
+	ctx           context.Context
```
Comment (CarlHembrough) [inline service/service.go]: This does not appear to be used? Consider removing it.

File: service/service.go (hunk 1 above)
Comment (CarlHembrough) [reply service/service.go]: Should we be doing this as part of this PR?

**Comment by DavidSubiros** [comment]
About the duplicates:
- I could reproduce the issue when the dependencies are running (via websysd dp-run.sh)
- If the dependencies are not running I only see one log per dependency per iteration.

File: service/service.go (hunk 1 above)
Comment (DavidSubiros) [reply service/service.go]: I think this should be part of another issue.
In this task I created the healthcheck client here (using NewAPIClient from dp-api-clients-go), and used it in the go-ns identity Handler. Before this change, we were creating the HTTP client in go-ns/identity

I imagine at some point we want to move the Handler out of go-ns/identity into dp-api-clients-go/identity, but I think this should be a separate task. What do you think?

**Comment by nshumoogum** [comment]
As David mentioned, the duplicates are to do with handling applications that do not have a `/health` endpoint yet, and instead api checker will call `/healthcheck` on not found response by API. This is expected until (like David has said) we update services to have a `/health` endpoint

File: ci/build.yml [hunk 3]
```diff
@@ -6,7 +6,7 @@ image_resource:
   type: docker-image
   source:
     repository: golang
-    tags: 1.13.7
+    tag: 1.13.7
```
Comment (nshumoogum) [inline ci/build.yml]: We can update this to 1.13.8, thanks

File: go.sum [hunk 4]
```diff
@@ -1,57 +1,63 @@
 cloud.google.com/go v0.26.0/go.mod h1:aQUYkXzVsufM+DwF1aE+0xfcU+56JwCaLick0ClmMTw=
```
Comment (nshumoogum) [inline go.sum]: Remember to run `go mod tidy` :)

File: service/service.go [hunk 5]
```diff
@@ -110,43 +94,42 @@ func Create(
...
 	d.server.HandleOSSignals = false
-	ctx, cancel := context.WithTimeout(context.Background(), d.shutdown)
 
-	// Start Healthcheck and server
 	d.healthCheck.Start(ctx)
-	d.run()
+	d.run(ctx)
 
 	select {
 	case err := <-d.errChan:
```
Comment (nshumoogum) [inline service/service.go]: Should this be triggering a shutdown?

Comment (DavidSubiros) [reply service/service.go]: good spot! I will log the error, but not shut down in this case.

Comment (DavidSubiros) [reply service/service.go]: Actually, if we don't want to shut down on this event, I can just remove this channel (the error is logged anyway in run func)

File: main.go [hunk 6]
```diff
@@ -95,29 +96,40 @@ func registerCheckers(ctx context.Context, hc *healthcheck.HealthCheck, isPublis
 	vc *vault.Client,
 	fc *filter.Client,
 	zc *health.Client,
-	s3 *s3client.S3) (err error) {
+	s3 *s3client.S3) error {
 
-	if err = hc.AddCheck("Dataset API", dc.Checker); err != nil {
+	anyError := false
```
Comment (nshumoogum) [inline main.go]: Maybe hasErrors?

**Pull request merged**